# Python Pandas

# 1 . Data Preparation

## 1.1 Philadelphia 2019

### Trip Data

In [1]:
# import all necessary packages
import pandas as pd
import numpy as np
import json
#import seaborn as sns
#import matplotlib as mlp
#import matplotlib.pyplot as plt
#from datetime import datetime

In [2]:
# import dataset
dfPhiladelphia = pd.read_csv("data/philadelphia_2019.csv")
dfPhiladelphia.head()

start_time             end_time  start_station_id  end_station_id  \
0  2019-01-01 00:19:00  2019-01-01 00:27:00              3049            3007   
1  2019-01-01 00:30:00  2019-01-01 00:37:00              3005            3007   
2  2019-01-01 00:52:00  2019-01-01 01:05:00              3166            3169   
3  2019-01-01 00:55:00  2019-01-01 01:04:00              3058            3103   
4  2019-01-01 01:05:00  2019-01-01 01:17:00              3182            3028   

   bike_id user_type    start_station_name  \
0    14495  Indego30       Foglietta Plaza   
1     5332  Day Pass     Welcome Park, NPS   
2    14623  Indego30  Frankford & Belgrade   
3    11706  Indego30      20th & Fairmount   
4    11039  Indego30         17th & Sansom   

                            end_station_name  
0                     11th & Pine, Kahn Park  
1                     11th & Pine, Kahn Park  
2                                 2nd & Race  
3  27th & Master, Athletic Recreation Center  
4                           4th & Bainbridge

In [3]:
dfPhiladelphia.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 744260 entries, 0 to 744259
Data columns (total 8 columns):
 #   Column              Non-Null Count   Dtype 
---  ------              --------------   ----- 
 0   start_time          744260 non-null  object
 1   end_time            744260 non-null  object
 2   start_station_id    744260 non-null  int64 
 3   end_station_id      744260 non-null  int64 
 4   bike_id             744260 non-null  int64 
 5   user_type           744225 non-null  object
 6   start_station_name  744260 non-null  object
 7   end_station_name    744260 non-null  object
dtypes: int64(3), object(5)
memory usage: 45.4+ MB


In [4]:
# Test if null values are existing in dfPhiladelphia
len(dfPhiladelphia)-len(dfPhiladelphia.dropna())

35

There are 35 rows with null values existing in dfPhiladelphia (but just in column user_type)

In [5]:
# Drop rows with null values
dfPhiladelphia.dropna(inplace=True)

In [6]:
# Test if duplicate rows are existing in dfPhiladelphia
len(dfPhiladelphia)-len(dfPhiladelphia.drop_duplicates())

196

There are 196 duplicates in dfPhiladelphia.
Since no bike (unique bike_id) can be rented more than once at the same time we will drop the duplicates

In [7]:
# Drop duplicate rows
dfPhiladelphia.drop_duplicates(inplace=True)

In [8]:
dfPhiladelphia.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 744029 entries, 0 to 744259
Data columns (total 8 columns):
 #   Column              Non-Null Count   Dtype 
---  ------              --------------   ----- 
 0   start_time          744029 non-null  object
 1   end_time            744029 non-null  object
 2   start_station_id    744029 non-null  int64 
 3   end_station_id      744029 non-null  int64 
 4   bike_id             744029 non-null  int64 
 5   user_type           744029 non-null  object
 6   start_station_name  744029 non-null  object
 7   end_station_name    744029 non-null  object
dtypes: int64(3), object(5)
memory usage: 51.1+ MB


In [9]:
# use the correct datatypes for the columns "start_time" and "end_time"
dfPhiladelphia['start_time'] = pd.to_datetime(dfPhiladelphia['start_time'])
dfPhiladelphia['end_time'] = pd.to_datetime(dfPhiladelphia['end_time'])

In [10]:
dfPhiladelphia.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 744029 entries, 0 to 744259
Data columns (total 8 columns):
 #   Column              Non-Null Count   Dtype         
---  ------              --------------   -----         
 0   start_time          744029 non-null  datetime64[ns]
 1   end_time            744029 non-null  datetime64[ns]
 2   start_station_id    744029 non-null  int64         
 3   end_station_id      744029 non-null  int64         
 4   bike_id             744029 non-null  int64         
 5   user_type           744029 non-null  object        
 6   start_station_name  744029 non-null  object        
 7   end_station_name    744029 non-null  object        
dtypes: datetime64[ns](2), int64(3), object(3)
memory usage: 51.1+ MB


In [11]:
# split date and time for start_time and end_time, but keep nonsplited for the weather data merge
dfPhiladelphia['start_time_for_weather_merge'] = dfPhiladelphia['start_time']
dfPhiladelphia['end_time_for_weather_merge'] = dfPhiladelphia['end_time']
dfPhiladelphia['start_date'] = dfPhiladelphia['start_time'].dt.date
dfPhiladelphia['start_time'] = dfPhiladelphia['start_time'].dt.time
dfPhiladelphia['end_date'] = dfPhiladelphia['end_time'].dt.date
dfPhiladelphia['end_time'] = dfPhiladelphia['end_time'].dt.time
dfPhiladelphia.head()

start_time  end_time  start_station_id  end_station_id  bike_id user_type  \
0   00:19:00  00:27:00              3049            3007    14495  Indego30   
1   00:30:00  00:37:00              3005            3007     5332  Day Pass   
2   00:52:00  01:05:00              3166            3169    14623  Indego30   
3   00:55:00  01:04:00              3058            3103    11706  Indego30   
4   01:05:00  01:17:00              3182            3028    11039  Indego30   

     start_station_name                           end_station_name  \
0       Foglietta Plaza                     11th & Pine, Kahn Park   
1     Welcome Park, NPS                     11th & Pine, Kahn Park   
2  Frankford & Belgrade                                 2nd & Race   
3      20th & Fairmount  27th & Master, Athletic Recreation Center   
4         17th & Sansom                           4th & Bainbridge   

  start_time_for_weather_merge end_time_for_weather_merge  start_date  \
0          2019-01-01 00:19:00        2019-01-01 00:27:00  2019-01-01   
1          2019-01-01 00:30:00        2019-01-01 00:37:00  2019-01-01   
2          2019-01-01 00:52:00        2019-01-01 01:05:00  2019-01-01   
3          2019-01-01 00:55:00        2019-01-01 01:04:00  2019-01-01   
4          2019-01-01 01:05:00        2019-01-01 01:17:00  2019-01-01   

     end_date  
0  2019-01-01  
1  2019-01-01  
2  2019-01-01  
3  2019-01-01  
4  2019-01-01

In [12]:
#maybe change datatype of station names and user type to string

In [13]:
# Is the feature user_type valuable for our task? What values are inside?
dfPhiladelphia["user_type"].unique()

array(['Indego30', 'Day Pass', 'Indego365', 'IndegoFlex', 'Walk-up'],
      dtype=object)

Drop features we won't need for all the other tasks
(bike_id?, start_station_name, end_station_name)

In [14]:
#...

### Station Data

In this section we will:
 - import the coordination data (latitude & longitude) for each station_id
 - merge latitude & longitude into dfPhiladelphia (merge over station_id)

In [15]:
# load station data using Python JSON module
with open('data/stations.json','r') as f:
    stations = json.loads(f.read())
# Flatten data
dfStations = pd.json_normalize(stations, record_path =['features'])
dfStations

type   geometry.coordinates geometry.type  properties.id  \
0    Feature  [-75.16374, 39.95378]         Point           3004   
1    Feature  [-75.14403, 39.94733]         Point           3005   
2    Feature   [-75.20311, 39.9522]         Point           3006   
3    Feature  [-75.15993, 39.94517]         Point           3007   
4    Feature  [-75.15067, 39.98081]         Point           3008   
..       ...                    ...           ...            ...   
182  Feature  [-75.16027, 39.95028]         Point           3295   
183  Feature  [-75.16758, 39.95134]         Point           3296   
184  Feature  [-75.20691, 39.94717]         Point           3298   
185  Feature  [-75.15698, 39.94332]         Point           3299   
186  Feature  [-75.16287, 39.95511]         Point           3300   

                       properties.name properties.coordinates  \
0    Municipal Services Building Plaza  [-75.16374, 39.95378]   
1                    Welcome Park, NPS  [-75.14403, 39.94733]   
2                        40th & Spruce   [-75.20311, 39.9522]   
3               11th & Pine, Kahn Park  [-75.15993, 39.94517]   
4            Temple University Station  [-75.15067, 39.98081]   
..                                 ...                    ...   
182                    12th & Chestnut  [-75.16027, 39.95028]   
183                    16th & Chestnut  [-75.16758, 39.95134]   
184                    42nd & Woodland  [-75.20691, 39.94717]   
185                      9th & Lombard  [-75.15698, 39.94332]   
186                     Broad & Cherry  [-75.16287, 39.95511]   

     properties.totalDocks  properties.docksAvailable  \
0                       30                          7   
1                       13                          4   
2                       17                         13   
3                       20                         16   
4                       17                          9   
..                     ...                        ...   
182                     21                         12   
183                     21                         16   
184                     21                         17   
185                     22                         14   
186                     13                          7   

     properties.bikesAvailable  properties.classicBikesAvailable  ...  \
0                           22                                18  ...   
1                            8                                 5  ...   
2                            3                                 0  ...   
3                            4                                 3  ...   
4                            8                                 2  ...   
..                         ...                               ...  ...   
182                          9                                 4  ...   
183                          4                                 1  ...   
184                          4                                 0  ...   
185                          8                                 5  ...   
186                          6                                 3  ...   

     properties.isEventBased  properties.isVirtual  properties.kioskId  \
0                      False                 False                3004   
1                      False                 False                3005   
2                      False                 False                3006   
3                      False                 False                3007   
4                      False                 False                3008   
..                       ...                   ...                 ...   
182                    False                 False                3295   
183                    False                 False                3296   
184                    False                 False                3298   
185                    False                 False                3299   
186                    False      

In [16]:
dfStations = pd.DataFrame(dfStations, columns=['properties.id','geometry.coordinates'])
dfStations

properties.id   geometry.coordinates
0             3004  [-75.16374, 39.95378]
1             3005  [-75.14403, 39.94733]
2             3006   [-75.20311, 39.9522]
3             3007  [-75.15993, 39.94517]
4             3008  [-75.15067, 39.98081]
..             ...                    ...
182           3295  [-75.16027, 39.95028]
183           3296  [-75.16758, 39.95134]
184           3298  [-75.20691, 39.94717]
185           3299  [-75.15698, 39.94332]
186           3300  [-75.16287, 39.95511]

[187 rows x 2 columns]

In [17]:
dfStations.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 187 entries, 0 to 186
Data columns (total 2 columns):
 #   Column                Non-Null Count  Dtype 
---  ------                --------------  ----- 
 0   properties.id         187 non-null    int64 
 1   geometry.coordinates  187 non-null    object
dtypes: int64(1), object(1)
memory usage: 3.0+ KB


In [18]:
#
dfStations['geometry.coordinates'] = dfStations['geometry.coordinates'].astype('string')

In [19]:
dfStations.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 187 entries, 0 to 186
Data columns (total 2 columns):
 #   Column                Non-Null Count  Dtype 
---  ------                --------------  ----- 
 0   properties.id         187 non-null    int64 
 1   geometry.coordinates  187 non-null    string
dtypes: int64(1), string(1)
memory usage: 3.0 KB


In [20]:
#...

### Combine Station Data with Philadelphia Data

In this section we will:
 - add the latitude and longitude data from dfStations to dfPhiladelphia
 - (Remember to drop null values, bbecause there are new stations which didn't existed in 2019)

In [21]:
#...

## 1.2 Weatherdata

In this section we will:
 - import the weather data
 - prepare the weather data
 - merge the weather data into dfPhiladelphia (merge over start_time_for_weather_merge)
 - after merge remember to drop start_time_for_weather_merge